**Windows** [Build 1147](http://brontes.lab.eng.brq.redhat.com/viewLog.html?buildId=38070&buildTypeId=WF_PullRequest_Windows) outcome was **FAILURE** using a merge of a376d99cb4894be68710fd4ebe09be849d881d87
Summary: Tests failed: 2, passed: 2931, ignored: 314 Build time: 0:47:11
### Build problems:
Failed tests detected

### Failed tests
```
org.jboss.as.test.clustering.xsite.XSiteSimpleTestCase.testPutRelayedToBackups: java.lang.AssertionError: expected:<200> but was:<500>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)

org.jboss.as.test.manualmode.ejb.client.cluster.EJBClientClusterConfigurationTestCase.testServerToServerClusterFormation: javax.ejb.EJBException: java.lang.IllegalStateException: EJBCLIENT000029: No cluster context available for cluster named ejb
	at org.jboss.as.ejb3.tx.CMTTxInterceptor.handleExceptionInOurTx(CMTTxInterceptor.java:188)
	at org.jboss.as.ejb3.tx.CMTTxInterceptor.invokeInOurTx(CMTTxInterceptor.java:279)

```

--- a/ejb3/src/main/java/org/jboss/as/ejb3/component/interceptors/AsyncFutureInterceptorFactory.java
+++ b/ejb3/src/main/java/org/jboss/as/ejb3/component/interceptors/AsyncFutureInterceptorFactory.java
@@ -28,10 +28,12 @@ import java.security.PrivilegedAction;
 import org.jboss.as.ee.component.Component;
 import org.jboss.as.ee.component.interceptors.InvocationType;
 import org.jboss.as.ejb3.component.session.SessionBeanComponent;
+import org.jboss.as.security.remoting.RemotingContext;
 import org.jboss.invocation.Interceptor;
 import org.jboss.invocation.InterceptorContext;
 import org.jboss.invocation.InterceptorFactory;
 import org.jboss.invocation.InterceptorFactoryContext;
+import org.jboss.remoting3.Connection;
 import org.jboss.security.SecurityContext;
 import org.jboss.security.SecurityContextAssociation;
 import org.jboss.security.plugins.JBossSecurityContext;
@@ -89,14 +91,20 @@ public final class AsyncFutureInterceptorFactory implements InterceptorFactory {
                     // we can't do anything if it isn't a JBossSecurityContext so just use the original one
                     clonedSecurityContext = securityContext;
                 }
+                final Connection remoteConnection = getConnection();
                 final AsyncInvocationTask task = new AsyncInvocationTask(flag) {
                     @Override
                     protected Object runInvocation() throws Exception {
                         setSecurityContextOnAssociation(clonedSecurityContext);
+                        setConnection(remoteConnection);
                         try {
                             return asyncInterceptorContext.proceed();
                         } finally {
-                            clearSecurityContextOnAssociation();
+                            try {
+                                clearSecurityContextOnAssociation();
+                            } finally {
+                                clearConnection();
+                            }
                         }
                     }
                 };
@@ -115,6 +123,46 @@ public final class AsyncFutureInterceptorFactory implements InterceptorFactory {
                 return task;
             }
         };
+    }
+
+    private void setConnection(final Connection remoteConnection) {
+        if (WildFlySecurityManager.isChecking()) {
+            WildFlySecurityManager.doUnchecked(new PrivilegedAction<Void>() {
+                @Override
+                public Void run() {
+                    RemotingContext.setConnection(remoteConnection);
+                    return null;
+                }
+            });
+        } else {
+            RemotingContext.setConnection(remoteConnection);
+        }
+    }
+
+    private void clearConnection() {
+        if (WildFlySecurityManager.isChecking()) {
+            WildFlySecurityManager.doUnchecked(new PrivilegedAction<Void>() {
+                @Override
+                public Void run() {
+                    RemotingContext.clear();
+                    return null;
+                }
+            });
+        } else {
+            RemotingContext.clear();
+        }
+    }
+    private Connection getConnection() {
+        if(WildFlySecurityManager.isChecking()) {
+            return WildFlySecurityManager.doUnchecked(new PrivilegedAction<Connection>() {
+                @Override
+                public Connection run() {
+                    return RemotingContext.getConnection();
+                }
+            });
+        } else {
+            return RemotingContext.getConnection();
+        }
     }
 
     private static void setSecurityContextOnAssociation(final SecurityContext sc) {

